cpp/src/traversal/mg/frontier_expand_kernels.cuh

```diff
@@ -0,0 +1,292 @@
+    block_frontier[i + thread_write_offset] = thread_frontier[i];
+  }
+  __syncthreads();
+
+  // If the total number of frontiers for this block is 0 then return
+  if (*block_frontier_count == 0) { return; }
+
+  // Find out where to write the block frontier to global memory
+  if (threadIdx.x == 0) {
+    *block_write_offset = cugraph::detail::traversal::atomicAdd(
+      output_frontier_count, static_cast<edge_t>(*block_frontier_count));
+  }
+  __syncthreads();
+
+  // Write block frontier to global memory
+  for (int i = threadIdx.x; i < (*block_frontier_count); i += blockDim.x) {
+    output_frontier[(*block_write_offset) + i] = block_frontier[i];
+  }
+}
+
+template <int BlockSize, int EdgesPerThread, typename vertex_t, typename edge_t, typename Operator>
+__global__ void block_per_vertex(edge_t *offsets,
+                                 vertex_t *indices,
+                                 vertex_t *input_frontier,
+                                 vertex_t input_frontier_count,
+                                 vertex_t vertex_begin,
+                                 vertex_t *output_frontier,
+                                 edge_t *output_frontier_count,
+                                 Operator op)
+{
+  if (blockIdx.x >= input_frontier_count) { return; }
+
+  __shared__ edge_t block_write_offset;
+  __shared__ vertex_t block_frontier[BlockSize * EdgesPerThread];
+  __shared__ int block_frontier_count;
+  vertex_t thread_frontier[EdgesPerThread];
+
+  vertex_t source        = input_frontier[blockIdx.x];
+  edge_t beg_edge_offset = offsets[source];
+  edge_t end_edge_offset = offsets[source + 1];
+
+  edge_t edge_offset = threadIdx.x + beg_edge_offset;
+  int num_iter       = (end_edge_offset - beg_edge_offset + BlockSize - 1) / BlockSize;
+
+  int thread_frontier_count = 0;
+  for (int i = 0; i < num_iter; ++i) {
+    if (edge_offset < end_edge_offset) {
+      vertex_t destination = indices[edge_offset];
+      // If operator returns true then add to local frontier
+      if (op(source + vertex_begin, destination)) {
+        thread_frontier[thread_frontier_count++] = destination;
+      }
+    }
+    bool is_last_iter = (i == (num_iter - 1));
+    bool is_nth_iter  = (i % EdgesPerThread == 0);
+    // Write to frontier every EdgesPerThread iterations
+    // Or if it is the last iteration of the for loop
+    if (is_nth_iter || is_last_iter) {
+      write_to_frontier(thread_frontier,
+                        thread_frontier_count,
+                        block_frontier,
+                        &block_frontier_count,
+                        output_frontier,
+                        &block_write_offset,
+                        output_frontier_count);
+      thread_frontier_count = 0;
+    }
+    edge_offset += blockDim.x;
+  }
+}
+
+template <int BlockSize, int EdgesPerThread, typename vertex_t, typename edge_t, typename Operator>
+__global__ void kernel_per_vertex(edge_t *offsets,
```

Comment:
Please consider using (East-) `const` qualifiers for read-only pointer parameters (like `input_frontier`, I think)